### PR TITLE
Iml vendor scraper and us_nj_bergen

### PR DESCRIPTION
**Pull request by ohinds**

### CODE REVIEW

File: recidiviz/common/constants/bond.py [hunk 1]
```diff
@@ -52,6 +52,7 @@ def _get_default_map():
     'NO BOND': BondType.NO_BOND,
     'NO BOND ALLOWED': BondType.NO_BOND,
     'NONE SET': BondType.NO_BOND,
+    'RELEASED ON OWN RECOGNIZANCE': BondType.NO_BOND,
```
Comment (jessex) [inline recidiviz/common/constants/bond.py]: Just for edification, this is what ROR below refers to as well?

File: recidiviz/ingest/scraper.py [hunk 2]
```diff
@@ -224,21 +224,27 @@ def resume_scrape(self, scrape_type):
         self.add_task(self.get_initial_task(), params)
 
     @staticmethod
-    def fetch_page(url, post_data=None, json_data=None):
+    def fetch_page(url, post_data=None, json_data=None, extra_headers=None):
```
Comment (jessex) [inline recidiviz/ingest/scraper.py]: I think there's a merge conflict here. `scraper.fetch_page` now has `headers` as its second argument, which you should be able to use.

File: recidiviz/common/constants/charge.py [hunk 3]
```diff
@@ -77,9 +78,15 @@ def _get_default_map():
     '1ST': ChargeDegree.FIRST,
     '2ND': ChargeDegree.SECOND,
     '3RD': ChargeDegree.THIRD,
+    '4TH': ChargeDegree.FORTH,
```
Comment (arian487) [inline recidiviz/common/constants/charge.py]: Fourth?

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+            os.path.dirname(__file__), key_mapping_file)
+
+        bond_key_mapping_file = 'iml_bond.yaml'
+        self.bond_yaml_file = os.path.join(
+            os.path.dirname(__file__), bond_key_mapping_file)
+
+        super(ImlScraper, self).__init__(region_name)
+
+        self._base_endpoint = self.get_region().base_url
+        self._num_people_per_search = 30
```
Comment (jessex) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: Let's extract this number out to a constant in the file for visibility.

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+        """Parses the number of people out of the content.
+
+        Args:
+            content: An lxml html tree.
+
+        Returns:
+            The total number of people to be scraped.
+
+        """
+        matches = re.findall(b'([0-9]+?) results', html.tostring(content))
```
Comment (jessex) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: There's no CSS selector or anything to look for? I could envision, e.g. some odd SEO-related update, adding "X results" somewhere in the skeleton of the page.

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+        if len(matches) != 2:
```
Comment (jessex) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: Why do we expect two results? What's the second one? This relates to my comment one line up about a more resilient check.

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+from recidiviz.ingest.extractor.html_data_extractor import HtmlDataExtractor
+
+
+class ImlScraper(BaseScraper):
+    """Scraper for counties using Iml."""
+
+    def __init__(self, region_name):
+        key_mapping_file = 'iml.yaml'
+        self.yaml_file = os.path.join(
+            os.path.dirname(__file__), key_mapping_file)
```
Comment (arian487) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: I would bake in the option to have a vendor scraper pass in its own yaml file.  I don't know if its possible, but I can envision that some regions can have different fields populated even amongst vendor scrapers (this is true for archonix for example).  You can use jailtracker and us_vt as a model

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 4 above)
Comment (arian487) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: Add this as a class constant instead of a member instance variable.

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+
+        Returns:
+            a dictionary with data needed to get the next page of search
+            results.
+
+        """
+
+        # Get the last result index on the previous page.
+        # pylint:disable=anomalous-backslash-in-string
+        matches = re.findall(b'Showing\s*[0-9]+\s*to\s*([0-9]+)\s*of',
```
Comment (jessex) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: Same comments as above about looking for specific elements or classes instead of doing a big regex search, and the number of matches thereof.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 5 above)
Comment (arian487) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: I don't like the idea of parsing the content back to raw html, html parsing can be an expensive operation and that is somewthing I'd like to avoid when at scale.  I'd rather have a flag that you can pass down through the params that tells fetch content to return us parsed html or not.

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+        Args:
+            content: An lxml html tree.
+
+        Returns:
+            The total number of people to be scraped.
+
+        """
+        matches = re.findall(b'([0-9]+?) results', html.tostring(content))
+        if len(matches) != 2:
+            logging.error('Did not find exactly two matches when parsing the '
```
Comment (arian487) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: Can you add a comment here on why we always expect two?  Does this always hold true?

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+
+            sysID = int(ids[1])
+            imgSysID = int(ids[2])
+
+            # Get the person IDs for the DB.
+            td_as = row.xpath('td/a')
+
+            if len(td_as) != 4:
+                logging.error('Did not find exactly 4 numbers in the row '
+                              'click function call, as we should have.')
```
Comment (jessex) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: Is "row click function call" accurate here, or a copy+paste boo boo?

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+
+        Args:
+            content: An lxml html tree.
+            params: dict of parameters passed from the last scrape session.
+
+        Returns:
+            A list of params containing endpoint and task_type at minimum.
+        """
+        task_type = params.get('task_type', self.get_initial_task_type())
+        params_list = []
```
Comment (jessex) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: This is an example of Hungarian Notation and it's generally frowned upon, i.e. we shouldn't embed types in the names of our variables if we can avoid it. In this case, something like `task_params` may be more appropriate.

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+        Args:
+            content: An lxml html tree.
+            params: dict of parameters passed from the last scrape session.
+            ingest_info: The IngestInfo object to populate
+        """
+        data_extractor = HtmlDataExtractor(self.yaml_file)
+        ingest_info = data_extractor.extract_and_populate_data(content,
+                                                               ingest_info)
+        if len(ingest_info.person) != 1:
+            logging.error("Data extractor didn't find exactly one person as "
```
Comment (jessex) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: In other cases of error-level log statements in this PR, we short-circuit. But here we silently ignore the additional persons. Should we be ignoring them instead of short-circuiting? Or, at least, should we include in the log statement that we'll ignore the extra persons?

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+                          "it should have")
+
+        person = ingest_info.person[0]
```
Comment (jessex) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: What if `ingest_info.person` is empty?

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+        person.person_id = params['person_id']
+        person.booking[0].booking_id = params['booking_id']
```
Comment (jessex) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: Is `person.booking` guaranteed to be non-empty?

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+
+        if person.race.startswith('HISPANIC'):
+            person.ethnicity = 'HISPANIC'
+            person.race = ' '.join(person.race.split()[1:])
+
+        # Find the bonds and take their data
+        bond_tab = [t for t in content.xpath('//table')
```
Comment (jessex) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: This list comprehension is guaranteed to be non-empty? There's definitely a table with the text "Bond Information"?

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 5 above)
Comment (jamwalla) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: This is handled now in #637 and #650 if the request has `params.['response_type'] = constants.TEXT_RESPONSE_TYPE`

Comment (jamwalla) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: could we use the `<form/>` elements at the top of the page instead? Seems like we could use the `nextSearch` number until it reaches `lastSearch`

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 16]
```diff
@@ -0,0 +1,349 @@
...
+                                                               ingest_info)
+        if len(ingest_info.person) != 1:
+            logging.error("Data extractor didn't find exactly one person as "
+                          "it should have")
+
+        person = ingest_info.person[0]
+        person.person_id = params['person_id']
+        person.booking[0].booking_id = params['booking_id']
+
+        if person.race.startswith('HISPANIC'):
```
Comment (jamwalla) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: Do you think this is something to try to handle in the converter? I could see either way.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 15 above)
Comment (jamwalla) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: nit- does it work to replace `html.tostring(t)` with `t.text_content()`?

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+        bond_ingest_info = bond_extractor.extract_and_populate_data(bond_tab,
+                                                                    None)
+
+        # Connect bonds and charges, and fill in an ICE hold when we
+        # see it in the charge name.
+        for charge in person.booking[0].charge:
+            if charge.name == 'IMMIGRATION DETAINEE':
+                person.booking[0].hold = 'ICE'
+
+            if not bond_ingest_info.person:
```
Comment (jamwalla) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: this check should be above the for loop i think

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 18]
```diff
@@ -0,0 +1,349 @@
...
+
+        if person.race.startswith('HISPANIC'):
+            person.ethnicity = 'HISPANIC'
+            person.race = ' '.join(person.race.split()[1:])
+
+        # Find the bonds and take their data
+        bond_tab = [t for t in content.xpath('//table')
+                    if b'Bond Information' in html.tostring(t)][0]
+        bond_extractor = HtmlDataExtractor(self.bond_yaml_file)
+        bond_ingest_info = bond_extractor.extract_and_populate_data(bond_tab,
```
Comment (jamwalla) [inline recidiviz/ingest/vendors/iml/iml_scraper.py]: is it worth checking also if, after looping through the charges, there are any bonds from the table that did not match to any charge?

File: recidiviz/persistence/database/migrations/versions/3bf06134321e_add_chargedegree_forth.py [hunk 19]
```diff
@@ -0,0 +1,45 @@
+"""add_ChargeDegree_FORTH
```
Comment (zdg2102) [inline recidiviz/persistence/database/migrations/versions/3bf06134321e_add_chargedegree_forth.py]: Can we break this out? I'd prefer to have the enum change and migration go in as its own PR first, rather than being lumped in with all these other changes

File: recidiviz/common/constants/bond.py (hunk 1 above)
Comment (ohinds) [reply recidiviz/common/constants/bond.py]: yup!

File: recidiviz/ingest/scraper.py (hunk 2 above)
Comment (ohinds) [reply recidiviz/ingest/scraper.py]: Great, I'll pull that in.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 4 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Done

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 5 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: This is moot given the rework to use xpaths instead of res.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 9 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Same as above.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 10 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Copy/paste error. Fixed.

File: recidiviz/common/constants/charge.py (hunk 3 above)
Comment (ohinds) [reply recidiviz/common/constants/charge.py]: You canadian.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 11 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: I totally agree about hungarian notation in general, but I don't think this fits neatly into that. 'type' here doesn't refer to the datatype, but rather to the variety of task that's to be gotten. I guess I could call it task_variety, but that seems like just replacing this with a synonym.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 12 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Yup, I just missed the return. Fixed.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 13 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Same as above.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 14 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: I thought it was, but turns out no. Check added.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 15 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Yup, done.

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 20]
```diff
@@ -0,0 +1,349 @@
...
+        # Connect bonds and charges, and fill in an ICE hold when we
+        # see it in the charge name.
+        for charge in person.booking[0].charge:
+            if charge.name == 'IMMIGRATION DETAINEE':
+                person.booking[0].hold = 'ICE'
+
+            if not bond_ingest_info.person:
+                continue
+
+            this_charge = [c for c in
```
[the remark this answers is not in the document]
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: So for this one I left the list comprehension because I want to test if we got more than one bond for a single charge. That seem ok?

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 21]
```diff
@@ -0,0 +1,349 @@
...
+                           if c.case_number == charge.case_number]
+
+            if not this_charge:
+                continue
+
+            charge.bond = this_charge[0].bond
+            charge.judge_name = this_charge[0].judge_name
+
+            # Transfer the bond type to charge status, if we should.
+            _CHARGE_STATUSES = [
```
[the remark this answers is not in the document]
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: This was so I could have the minimal universe of replacements that I've seen right here. I'll document that, unless you'd rather I include all possibilities we could handle?

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 22]
```diff
@@ -0,0 +1,349 @@
...
+            'flow_action': 'searchbyname',
+            'quantity': self._num_people_per_search,
+            'systemUser_includereleasedinmate': 'N',
+            'systemUser_includereleasedinmate2': 'N',
+        }
+
+        return data
+
+    def _fetch_content(self, endpoint, post_data=None, json_data=None):
+        """Returns the page content. This is overriden from BaseScraper
```
[the remark this answers is not in the document]
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Done

File: recidiviz/persistence/database/migrations/versions/3bf06134321e_add_chargedegree_forth.py [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = '3bf06134321e'
+down_revision = '328448d2b4d2'
+branch_labels = None
+depends_on = None
+
+# Without 'FORTH'
```
[the remark this answers is not in the document]
Comment (ohinds) [reply recidiviz/persistence/database/migrations/versions/3bf06134321e_add_chargedegree_forth.py]: Done

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 16 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Could be? The check would need to be more robust than this, I imagine.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 17 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Yup.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 18 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: This will happen, because some bonds don't have a case number. I'm not sure what to do with those, do you have ideas?

File: recidiviz/ingest/vendors/iml/iml_scraper.py [hunk 24]
```diff
@@ -0,0 +1,349 @@
...
+specific navigation and data extraction. All counties that use IML should
+have their county-specific scrapers inherit from this class.
+
+Vendor-specific notes:
+    - No historical data is kept, only data on current people
+    - An initial request can be made to retrieve a page with all roster entries
+    - The person page contains the current booking info
+
+Background scraping procedure:
+    1. A get without session variables to find the total number of people to
```
[the remark this answers is not in the document]
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Done

File: recidiviz/persistence/database/migrations/versions/3bf06134321e_add_chargedegree_forth.py (hunk 19 above)
Comment (ohinds) [reply recidiviz/persistence/database/migrations/versions/3bf06134321e_add_chargedegree_forth.py]: Sure thing.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 11 above)
Comment (jessex) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Mm, that's fair, but I think the coincidence may be misleading. I think renaming to avoid the appearance of HN is worth the effort, personally, but feel free to ignore.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jessex) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: That's reasonable!

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (jessex) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Hm. Personally, it seems that not having an unexpected bug if we miss a possibility in the future and not having to define another collection of statuses outweighs the benefit of this sort of documentation. If that's something we want to document, I'd rather have a comment somewhere saying "We expect these statuses specifically..."

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 7 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Done

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 11 above)
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: #697

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (ohinds) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Ok, I'm on board. I made this change.

File: recidiviz/ingest/vendors/iml/iml_scraper.py (hunk 18 above)
Comment (jamwalla) [reply recidiviz/ingest/vendors/iml/iml_scraper.py]: Oh, totally. Maybe we could add the log there as a warning and see if we can come back to this? I filed #696 because this is the second vendor we have this problem with.
